src/main/java/com/sonymobile/jenkins/plugins/mq/mqnotifier/MQConnection.java

```diff
@@ -207,20 +240,115 @@ public void publish(JSONObject json) {
      * Sends messages from the message queue.
      */
     private void sendMessages() {
+        Channel channel = null;
+
         while (true) {
             try {
-                MessageData messageData = (MessageData)messageQueue.poll(SENDMESSAGE_TIMEOUT,
+                if (channel == null || !channel.isOpen()) {
+                    channel = createChannel();
+                    channel.confirmSelect();
+                    addMessageConfirmListener(channel);
+                    setShouldAddToQueue(true);
+                }
+                MessageData messageData = (MessageData) messageQueue.poll(SENDMESSAGE_TIMEOUT,
                                                                          TimeUnit.MILLISECONDS);
                 if (messageData != null) {
-                    getInstance().send(messageData.getExchange(), messageData.getRoutingKey(),
-                            messageData.getProps(), messageData.getBody());
+                    validateExchange(channel, messageData.getExchange());
+                    getInstance().sendOnChannel(messageData, channel);
                 }
             } catch (InterruptedException ie) {
                 LOGGER.info("sendMessages() poll() was interrupted: ", ie);
+            } catch (IOException ioe) {
+                LOGGER.error("error validating channel: ", ioe);
+            } catch (ChannelCreationException | MessageDeliveryException transientException) {
+                LOGGER.error(transientException.getMessage(), transientException.getCause());
+                setShouldAddToQueue(false);
+                try {
+                    Thread.sleep(CONNECTION_WAIT);
+                } catch (InterruptedException ie) {
+                    LOGGER.error("Thread.sleep() was interrupted", ie);
+                }
+            }
+        }
+    }
+
+
+    /**
+     * Enable or disable adding messages to the queue.
+     */
+    private synchronized void setShouldAddToQueue(boolean enabled) {
+        this.shouldAddToQueue = enabled;
+        notifyAll();
+    }
+
+    /**
+     * Wait until it's ok to add messages to the queue again
+     */
+    public synchronized void waitForQueue() {
+        while (!this.shouldAddToQueue) {
+            try {
+                wait();
+            } catch (InterruptedException e)  {
+                Thread.currentThread().interrupt();
+            }
+        }
+    }
+
+    /**
+     * Validate the exchange.
+     *
+     * @throws IOException if the channel is invalid for the exchange
+     */
+    private void validateExchange(Channel channel, String exchange) throws IOException {
```

Comment:
Is IOException really the correct exception to throw? IllegalArgumentException? Or perhaps you are trying to avoid RuntimeExceptions? If so, NullPointerException?

Reply:
Good question, channel.exchangeDeclarePassive(exchange) will actually throw an IOException if the exchange is not valid, so I thought I match that. Although, a NullPointerException would probably be more suitable?